File: data/src/main/java/com/attiladroid/data/DataProvider.java
package com.attiladroid.data;

import android.arch.lifecycle.LiveData;
import android.content.Context;

import com.attiladroid.data.db.AppDatabase;
import com.attiladroid.data.db.Dao.InfoDao;
import com.attiladroid.data.db.Dao.LinkDao;
import com.attiladroid.data.db.Dao.SetsDao;
import com.attiladroid.data.db.Dao.ThemeDao;
import com.attiladroid.data.db.Dao.WordDao;
import com.attiladroid.data.entities.DictionaryInfo;
import com.attiladroid.data.entities.Link;
import com.attiladroid.data.entities.Word;
import com.attiladroid.data.entities.Set;
import com.attiladroid.data.entities.Theme;

import java.util.List;

import io.reactivex.Flowable;


/**
 * Created by alex on 1/5/18.
 */

public class DataProvider {

    private InfoDao infoDao;
    private LinkDao linkDao;
    private SetsDao setsDao;
    private ThemeDao themeDao;
    private WordDao wordDao;

    public static DataProvider newInstance(Context context) {
        AppDatabase db = AppDatabase.Companion.newInstance(context);
        DataProvider provider = new DataProvider(db.infoDao(), db.linkDao(), db.setsDao(),
                db.themeDao(), db.wordDao());
        return provider;
    }

    private DataProvider(InfoDao infoDao, LinkDao linkDao, SetsDao setsDao, ThemeDao themeDao, WordDao wordDao) {
        this.infoDao = infoDao;
        this.linkDao = linkDao;
        this.setsDao = setsDao;
        this.themeDao = themeDao;
        this.wordDao = wordDao;
    }

    ///////////////////////////////////////////////////////////////////////////
    // Words methods
    ///////////////////////////////////////////////////////////////////////////

    public long insertWord(Word word) {
        return wordDao.insertWord(word);
    }

    public Word getWord(String word, String translation) {
        return wordDao.getWord(word, translation);
    }

    public Word getWordById(long id) {
        return wordDao.getWordById(id);
    }

    public void insertWords(Word... words) {
        wordDao.insertWords(words);
    }

    public void updateWords(Word... words) {
        wordDao.udateWords(words);
    }

    // TODO: 1/7/18 Temp methods. Delete them and replace 
    public void resetWordProgress(long id) {
        Word word = wordDao.getWordById(id);
        word.setStudied(0);
        word.setTrainedTw(0);
        word.setTrainedWt(0);

        wordDao.udateWords(word);
    }

    public void updateWordStatus() {
    }

    public List<Word> getDictionaryWords() {
        return wordDao.dictionaryWords();
    }

    public Flowable<List<Word>> getDictionaryWordsFlowable() {
        return wordDao.dictionaryWordsFlowable();
    }

    public List<Word> getTrainingWords(long setId) {
        return getTrainingWords(setId, Integer.MAX_VALUE);
    }

    public List<Word> getTrainingWords(long setId, int limit) {
        if (setId == -1) {
            return wordDao.dictionaryWords();
        } else return wordDao.getWordsBySetId(setId);
    }

    public List<Word> getVariants(long wordId, int limit, long setId) {
        if (setId == -1) {
            return wordDao.getVariants(wordId, limit);
        } else {
            return wordDao.getVarints(wordId, limit, setId);
        }
    }

    public Flowable<List<Word>> getWordsBySetIdRx(long setId) {
        return wordDao.getWordsBySetIdRx(setId);
    }

    public List<Word> getWordsBySetId(long setId) {
        return wordDao.getWordsBySetId(setId);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Sets methods
    ///////////////////////////////////////////////////////////////////////////

    public long insertSet(Set set) {
        return setsDao.insertSet(set);
    }

    public void insertSets(Set... sets) {
        setsDao.insertSets(sets);
    }

    public void updateSets(Set... sets) {
        setsDao.updateSets(sets);
    }

    public void switchSetStatus(Set set) {
        int status = set.getStatus() == DatabaseContract.Sets.IN_DICTIONARY ?
                DatabaseContract.Sets.OUT_OF_DICTIONARY : DatabaseContract.Sets.IN_DICTIONARY;
        set.setStatus(status);
        updateSetStatus(set);
    }

    public void updateSetStatus(Set set) {
        setsDao.updateSets(set);
        List<Word> words = wordDao.getWordsBySetId(set.getId());
        long time = System.currentTimeMillis();
        boolean updateTime = set.getStatus() == DatabaseContract.Sets.IN_DICTIONARY;
        for (Word word : words) {
            word.setStatus(set.getStatus());
            if (updateTime) word.setAddedTime(time);
        }
        Word[] wordsArray = words.toArray(new Word[0]);
        wordDao.udateWords(wordsArray);
    }

    public void resetSetProgress(Set set) {
        List<Word> words = wordDao.getWordsBySetId(set.getId());
        for(Word word : words) {
            word.resetProgress();
        }
        wordDao.udateWords(words.toArray(new Word[0]));
    }

    public List<Set> getAllSets() {

        return setsDao.allSets();
    }

    public LiveData<List<Set>> getAllSetsLd() {
        return setsDao.allSetsAsync();
    }

    public Flowable<Set> getSetById(long setId) {
        return setsDao.getSetById(setId);
    }

    // TODO: 1/25/18 make with string
    public List<Set> getSetsByTheme(String themeCode) {
        return setsDao.getSetsByTheme("%" + themeCode + "%");
    }

    ///////////////////////////////////////////////////////////////////////////
    // Links, themes and info
    ///////////////////////////////////////////////////////////////////////////

    public void insertLink(Link link) {
        linkDao.insertLinks(link);
    }

    public void insertLinks(Link... links) {
        linkDao.insertLinks(links);
    }

    public LiveData<DictionaryInfo> getDictionaryInfo() {
        return infoDao.getDictionaryInfo();
    }

    public long insertTheme(Theme theme) {
        return themeDao.insertTheme(theme);
    }

    public void insertThemes(Theme... themes) {
        themeDao.insertThemes(themes);
    }

    public LiveData<List<Theme>> getAllThemes() {
        return themeDao.allThemes();
    }
}

File: data/src/main/java/com/attiladroid/data/DataProvider.kt
package com.attiladroid.data

import android.arch.lifecycle.LiveData
import android.content.Context

import com.attiladroid.data.db.AppDatabase
import com.attiladroid.data.db.Dao.InfoDao
import com.attiladroid.data.db.Dao.LinkDao
import com.attiladroid.data.db.Dao.SetsDao
import com.attiladroid.data.db.Dao.ThemeDao
import com.attiladroid.data.db.Dao.WordDao
import com.attiladroid.data.entities.DictionaryInfo
import com.attiladroid.data.entities.Link
import com.attiladroid.data.entities.Word
import com.attiladroid.data.entities.Set
import com.attiladroid.data.entities.Theme

import io.reactivex.Flowable


/**
 * Created by alex on 1/5/18.
 */

class DataProvider private constructor(
        private val infoDao: InfoDao,
        private val linkDao: LinkDao,
        private val setsDao: SetsDao,
        private val themeDao: ThemeDao,
        private val wordDao: WordDao) {

    companion object {
        fun newInstance(context: Context): DataProvider {
            val db = AppDatabase.newInstance(context)
            return DataProvider(db.infoDao(), db.linkDao(), db.setsDao(),
                    db.themeDao(), db.wordDao())
        }
    }

    val dictionaryWords: List<Word>
        get() = wordDao.dictionaryWords()

    val dictionaryWordsFlowable: Flowable<List<Word>>
        get() = wordDao.dictionaryWordsFlowable()

    val allSets: List<Set>
        get() = setsDao.allSets()

    val allSetsLd: LiveData<List<Set>>
        get() = setsDao.allSetsAsync()

    val dictionaryInfo: LiveData<DictionaryInfo>
        get() = infoDao.dictionaryInfo

    val allThemes: LiveData<List<Theme>>
        get() = themeDao.allThemes()

    ///////////////////////////////////////////////////////////////////////////
    // Words methods
    ///////////////////////////////////////////////////////////////////////////

    fun insertWord(word: Word): Long {
        return wordDao.insertWord(word)
    }

    fun getWord(word: String, translation: String): Word {
        return wordDao.getWord(word, translation)
    }

    fun getWordById(id: Long): Word {
        return wordDao.getWordById(id)
    }

    fun insertWords(vararg words: Word) {
        wordDao.insertWords(*words)
    }

    fun updateWords(vararg words: Word) {
        wordDao.udateWords(*words)
    }

    // TODO: 1/7/18 Temp methods. Delete them and replace
    fun resetWordProgress(id: Long) {
        val word = wordDao.getWordById(id)
        word.studied = 0
        word.trainedTw = 0
        word.trainedWt = 0

        wordDao.udateWords(word)
    }

    fun updateWordStatus() {}

    @JvmOverloads
    fun getTrainingWords(setId: Long, limit: Int = Integer.MAX_VALUE): List<Word> {
        return if (setId == -1L) {
            wordDao.dictionaryWords()
        } else
            wordDao.getWordsBySetId(setId)
    }

    fun getVariants(wordId: Long, limit: Int, setId: Long): List<Word> {
        return if (setId == -1L) {
            wordDao.getVariants(wordId, limit)
        } else {
            wordDao.getVarints(wordId, limit, setId)
        }
    }

    fun getWordsBySetId(setId: Long): LiveData<MutableList<Word>> {
        return wordDao.getWordsBySetIdLiveData(setId)
    }

    ///////////////////////////////////////////////////////////////////////////
    // Sets methods
    ///////////////////////////////////////////////////////////////////////////

    fun insertSet(set: Set): Long {
        return setsDao.insertSet(set)
    }

    fun insertSets(vararg sets: Set) {
        setsDao.insertSets(*sets)
    }

    fun updateSets(vararg sets: Set) {
        setsDao.updateSets(*sets)
    }

    fun switchSetStatus(set: Set) {
        val status = if (set.status == DataContract.Sets.IN_DICTIONARY)
            DataContract.Sets.OUT_OF_DICTIONARY
        else
            DataContract.Sets.IN_DICTIONARY
        set.status = status
        updateSetStatus(set)
    }

    fun updateSetStatus(set: Set) {
        setsDao.updateSets(set)
        val words = wordDao.getWordsBySetId(set.id)
        val time = System.currentTimeMillis()
        val updateTime = set.status == DataContract.Sets.IN_DICTIONARY
        for (word in words) {
            word.status = set.status
            if (updateTime) word.addedTime = time
        }
        val wordsArray = words.toTypedArray()
        wordDao.udateWords(*wordsArray)
    }

    fun resetSetProgress(set: Set) {
        val words = wordDao.getWordsBySetId(set.id)
        for (word in words) {
            word.resetProgress()
        }
        wordDao.udateWords(*words.toTypedArray())
    }

    fun getSetById(setId: Long): LiveData<Set> {
        return setsDao.getSetById(setId)
    }

    // TODO: 1/25/18 make with string
    fun getSetsByTheme(themeCode: String): List<Set> {
        return setsDao.getSetsByTheme("%$themeCode%")
    }

    ///////////////////////////////////////////////////////////////////////////
    // Links, themes and info
    ///////////////////////////////////////////////////////////////////////////

    fun insertLink(link: Link) {
        linkDao.insertLinks(link)
    }

    fun insertLinks(vararg links: Link) {
        linkDao.insertLinks(*links)
    }

    fun insertTheme(theme: Theme): Long {
        return themeDao.insertTheme(theme)
    }

    fun insertThemes(vararg themes: Theme) {
        themeDao.insertThemes(*themes)
    }
}
